### PR TITLE
Adjustments to Normaliz 3.3.0

### PR DESCRIPTION
**Pull request by sebasguts**
* Added new cone properties, and thereby removed some C-functions which were used to access properties that can now be accessed via cone properties.
* Implemented signal handling to interrupt normaliz omputations
* Use the normaliz autotools build system
* Use Normaliz version 3.3.0
* Adjusted test cases

### CODE REVIEW

File: src/normaliz.cc [hunk 1]
```diff
@@ -617,21 +686,73 @@ static Obj _NmzConePropertyImpl(Obj cone, Obj prop)
...
+    case libnormaliz::ConeProperty::NoSymmetrization:
+    case libnormaliz::ConeProperty::BigInt:
+    case libnormaliz::ConeProperty::NoNestedTri:
+    case libnormaliz::ConeProperty::HSOP:
+        ErrorQuit("Cone property is input only", 0, 0);
 
     default:
-        // Case not handled. Should signal an error
-        break;
+        ErrorQuit("Unknown cone property", 0, 0);
```
Comment (fingolfin) [inline src/normaliz.cc]: Calling `ErrorQuit` is wrong! It'll `longjmp` out, bypassing the C++ `catch`, preventing any C++ locals from deconstructing etc.

As an alternative, you could throw an special exception, and modify FUNC_END to trigger ErrorQuit for that. Or just print the error message using `Pr` and return fail. Or come up with some other mechanism.

File: build-normaliz.sh [hunk 2]
```diff
@@ -41,7 +41,8 @@ if [ -f "$GAP_GMP/MAKE_CHECK_PASSED" -a  -f "$GAP_GMP/include/gmp.h" ]; then
     echo "GAP was built with its own GMP"
     if [ -f "$GAP_GMP/include/gmpxx.h" ]; then
         echo "GAP's GMP includes C++ support"
-        export GMP_DIR="$GAP_GMP"
+#         export GMP_DIR="$GAP_GMP"
```
Comment (fingolfin) [inline build-normaliz.sh]: Why is the above line commented out but kept around?

File: src/normaliz.cc [hunk 3]
```diff
@@ -284,6 +290,30 @@ static Obj NmzMatrixToGAP(const vector< vector<Integer> >& in)
...
+    size_t n;
+    M = NEW_PLIST(T_PLIST, m);
+    SET_LEN_PLIST(M, m);
+    for (size_t i = 0; i < m; ++i) {
+        n = in[i].size();
+        N = NEW_PLIST( T_PLIST, n );
+        SET_LEN_PLIST( N, n );
+        for( size_t j = 0; j < n; ++j ){
+            SET_ELM_PLIST(N, j+1, in[i][j] ? True : False );
+            CHANGED_BAG( N );
```
Comment (fingolfin) [inline src/normaliz.cc]: There is no need to call CHANGED_BAG in each loop iteration here, as you allocate no new memory, so no GC can happen (and anyway, `True` and `False` won't be garbage collected)

File: src/normaliz.cc [hunk 4]
```diff
@@ -284,6 +290,30 @@ static Obj NmzMatrixToGAP(const vector< vector<Integer> >& in)
...
+{
+    Obj M;
+    Obj N;
+    const size_t m = in.size();
+    size_t n;
+    M = NEW_PLIST(T_PLIST, m);
+    SET_LEN_PLIST(M, m);
+    for (size_t i = 0; i < m; ++i) {
+        n = in[i].size();
+        N = NEW_PLIST( T_PLIST, n );
```
Comment (fingolfin) [inline src/normaliz.cc]: You could also use BLISTs here... Might be more work, though. I have no idea what these boolean lists are used for, so i can't judge if it's worth it. But if they can be large, it might be something to pursue (but perhaps in another patch; still, could at least a TODO comment here, and/or file a TODO issue on the tracker.

File: src/normaliz.cc [hunk 5]
```diff
@@ -284,6 +290,30 @@ static Obj NmzMatrixToGAP(const vector< vector<Integer> >& in)
...
+            CHANGED_BAG( N );
+        }
+        SET_ELM_PLIST( M, i+1, N );
+        CHANGED_BAG( M );
+    }
+    CHANGED_BAG( M );
+    return M;
+}
+
+// TODO: HSOP
```
Comment (fingolfin) [inline src/normaliz.cc]: what does that comment mean?

File: src/normaliz.cc [hunk 6]
```diff
@@ -295,6 +325,19 @@ static Obj NmzHilbertSeriesToGAP(const libnormaliz::HilbertSeries& HS)
     return ret;
 }
 
+template<typename Integer>
+static Obj NmzWeightedEhrhartSeriesToPyList(const std::pair<libnormaliz::HilbertSeries,Integer>& HS)
```
Comment (fingolfin) [inline src/normaliz.cc]: What's a `PyList`? ;-)

Also, this function seems to be unused, so why is it added?

File: src/normaliz.cc [hunk 7]
```diff
@@ -506,14 +565,15 @@ static Obj _NmzConePropertyImpl(Obj cone, Obj prop)
...
-        return NmzHilbertQuasiPolynomialToGAP(C->getHilbertSeries());
-    }
+//     if (string(CSTR_STRING(prop)) == string("HilbertQuasiPolynomial")) {
+//         C->compute(ConeProperties(libnormaliz::ConeProperty::HilbertSeries));
+//         return NmzHilbertQuasiPolynomialToGAP(C->getHilbertSeries());
+//     }
 
     libnormaliz::ConeProperty::Enum p = libnormaliz::toConeProperty(CSTR_STRING(prop));
-
+    current_interpreter_sigint_handler = signal( SIGINT, signal_handler );
```
Comment (fingolfin) [inline src/normaliz.cc]: oh, so you query the signal handler again here anyway? Now I am confused... perhaps rethink and/or explain how you deal with the signal handler switching?

File: src/normaliz.cc [hunk 8]
```diff
@@ -735,94 +837,16 @@ static Obj _NmzBasisChangeIntern(Obj cone)
...
-    FUNC_END
-}
+// Obj _NmzBasisChange(Obj self, Obj cone)
+// {
+//     FUNC_BEGIN
+//     if (!IS_CONE(cone))
+//         ErrorQuit("<cone> must be a Normaliz cone", 0, 0);
+//     return _NmzBasisChangeIntern<mpz_class>(cone);
+//     FUNC_END
+// }
```
Comment (fingolfin) [inline src/normaliz.cc]: Why is this commented out code kept around?

File: src/normaliz.h [hunk 9]
```diff
@@ -31,25 +31,37 @@ extern "C" {
 #include "libnormaliz/cone.h"
 #include <assert.h>
 
+#include <csignal>
+using std::signal;
+
+typedef void (*sighandler_t)(int);
```
Comment (fingolfin) [inline src/normaliz.h]: Golden rule for C++: never put a `using` into a header file :-).

In this case, though, I'd not even put the include nor the typedef into the header, they can and should both go into the .c file.

Comment (fingolfin) [inline src/normaliz.h]: ... unless you perhaps plan to split the source into multiple files... but even then I am sceptical it's needed in the header (and the `using` definitely shouldn't be in there -- indeed, why not use use `signal.h` directly?

Comment (fingolfin) [inline src/normaliz.h]: (Right now, technically speaking, there isn't really any reason why we have that header file... (it's mostly my fault it exists, I guess)

File: build-normaliz.sh [hunk 10]
```diff
@@ -95,8 +94,14 @@ if [ $GAParch_abi = "32-bit" ]; then
...
+
 PREFIX="$PWD/DST"
-cd BUILD
-cmake -DCMAKE_INSTALL_PREFIX:PATH="$PREFIX" ../source
+./bootstrap.sh
+if [ "x$GMP_FLAG" != "x" ]; then
+  ./configure --prefix=$PREFIX --with-gmp=$GMP_FLAG
+else
+  ./configure --prefix=$PREFIX
+fi
```
Comment (fingolfin) [inline build-normaliz.sh]: It would be helpful if one could pass additional arguments to `./configure` -- e.g. right now, I cannot tell it where to find my version of boost.

For the `cmake` build system, I could do this via an environment variable, but now I can't (or perhaps I can by modifying CPPFLAGS, LDFLAGS, CXXFLAGS suitably, but that's more cumbersome)

Comment (fingolfin) [inline build-normaliz.sh]: .... aha, but upon trying, it seems `Normaliz` does not actually have a `--with-boost=PREFIX` switch, so I'll have to mess with the FLAGS vars anyway.

**Comment by fingolfin** [comment]
On a meta-level, the commit structuring is not very helpful; e.g. there are three commits in a row adding signal handling, which could be squashed into one; as could all those commits which tweak the examples; or the two commits which modify `build-normaliz.sh`.

And the commit which adds `UnitGroupIndex` to the kernel extension comes *after* the commit which  add it to `lib/cone_property_wrappers.{gi,gd}`...

File: lib/cone_property_wrappers.gi [hunk 11]
```diff
@@ -1,63 +1,90 @@
 ## This is an automatically generated file
```
Comment (fingolfin) [inline lib/cone_property_wrappers.gi]: This line is actually completely untrue ("out-of-date"), isn't it? Same for its counterpart in the .gd file.

File: lib/cone_property_wrappers.gd [hunk 12]
```diff
@@ -231,3 +231,94 @@ DeclareGlobalFunction( "NmzTriangulationSize" );
...
+DeclareGlobalFunction( "NmzEquations" );
+
+#! @Arguments cone
+#! @Returns a matrix whose rows represent the congruences
+#! @Description
+#! Together with the support hyperplanes and the equations it describes the
+#! lattice points of the cone.
+#! <P/>
+#! This is part of the cone property <Q>Sublattice</Q>.
+DeclareGlobalFunction( "NmzCongruences" );
```
Comment (fingolfin) [inline lib/cone_property_wrappers.gd]: You removed `NmzCongruences` from the kernel and added no replacement function.

We really should have a test case for every single `Nmz*` function, to make sure they all actually exist, and do not crash ;-) (this is not something I am requesting for this PR, mind you, just a general thought; perhaps we should file an issue on this as reminder).

**Comment by fingolfin** [comment]
Two more things regarding signal handler: First off, consider using `sigaction` instead of `signal`. Secondly, you use the custom signal handler only for the two calls to Cone::compute(). Well, then don't put the signal handling stuff into FUNC_END at all! Instead, replace this code
```C++
    Cone<mpz_class>* C = GET_CONE<mpz_class>(cone);
    current_interpreter_sigint_handler = signal( SIGINT, signal_handler );
    ConeProperties notComputed = C->compute(propsToCompute);
    signal( SIGINT, current_interpreter_sigint_handler );
```
by this:

```C++
// Use RAII to set and reset signal handlers, even if exceptions are thrown
class TempSignalHandler {
    sighandler_t oldHandler;
public:
    SignalGuard(sighandler_t newHandler) {
        oldHandler = signal( SIGINT, newHandler );
    }
    ~SignalGuard() {
        signal( SIGINT, oldHandler );
    }
};

static ConeProperties compute(Cone<mpz_class> *C, ConeProperties propsToCompute) {
    TempSignalHandler tmp(signal_handler);
    return C->compute(propsToCompute);
}
...
    Cone<mpz_class>* C = GET_CONE<mpz_class>(cone);
    ConeProperties notComputed = compute(C, propsToCompute);
```
Of course you can also do this without the `compute` helper, but then you need to pay attention for when the scope ends.

File: src/normaliz.cc (hunk 1 above)
Comment (fingolfin) [inline src/normaliz.cc]: A simple way to resolve this `ErrorQuit` thing is to merge `_NmzConePropertyImpl` into `_NmzConeProperty` -- it was only ever separate in order to allow for different `Integer` types. Since this is no longer needed, we can get rid of that.

Once that is done, you can move the `FUNC_END` to *before* the big switch/case. Of course that is slightly risky, as the switch/case contains some more code that could throw.

So then an alternative is to keep the `FUNC_END` near the end, but set some error flags suitably in the switch/case, and check them after `FUNC_END`, then trigger the right `ErrorQuit` messages there.

Comment (fingolfin) [inline src/normaliz.cc]: Also, if you do errors here now (which I think is a good idea), then there should perhaps also be an error message for `StanleyDec`

File: src/normaliz.cc [hunk 13]
```diff
@@ -617,21 +686,73 @@ static Obj _NmzConePropertyImpl(Obj cone, Obj prop)
...
+    
     case libnormaliz::ConeProperty::Sublattice:
-        C->compute(p);
-        return C->isComputed(p) ? True : False;
+        return _NmzBasisChangeIntern(C);
+    
+    case libnormaliz::ConeProperty::IntegerHull:
+    {   
+        Cone<Integer>* hull = new Cone<Integer>( C->getIntegerHullCone() );
+        return NewCone( hull );
```
Comment (fingolfin) [inline src/normaliz.cc]: This code really should have a comment that explains why a copy of the integer hull cone is made (I assume for tracking-of-ownership reasons: The integer cone is "owned" by the cone C, so we must not wrap it directly into a GAP object). Anyway, this is the kind of reasoning that should be in a comment. (And if my assumption is wrong, then that's an even stronger argument for the need of a comment ;-)

File: src/normaliz.cc [hunk 14]
```diff
@@ -617,21 +686,73 @@ static Obj _NmzConePropertyImpl(Obj cone, Obj prop)
...
+    {   
+        Cone<Integer>* hull = new Cone<Integer>( C->getIntegerHullCone() );
+        return NewCone( hull ); 
+    }
+    
+    case libnormaliz::ConeProperty::HilbertQuasiPolynomial:
+        return NmzHilbertQuasiPolynomialToGAP(C->getHilbertSeries());
+    
+    case libnormaliz::ConeProperty::WeightedEhrhartQuasiPolynomial:
+        return NmzWeightedEhrhartQuasiPolynomialToGAP(C->getIntData());
```
Comment (fingolfin) [inline src/normaliz.cc]: As far as I can tell, this case is never explicitly tested...

So I guess that's one more request for "tests for all the new stuff would be really awesome."

File: build-normaliz.sh (hunk 2 above)
Comment (sebasguts) [inline build-normaliz.sh]: Removed in e80e0ec1d2f707907ff520555769aefeba9895dc

File: build-normaliz.sh (hunk 10 above)
Comment (sebasguts) [inline build-normaliz.sh]: Unfortunately no. I nevertheless added the possibility to pass flags to the configure script.

File: lib/cone_property_wrappers.gd (hunk 12 above)
Comment (sebasguts) [inline lib/cone_property_wrappers.gd]: Test cases should be there, yes. The reason I removed the kernel function is because `Congruences` is now a cone property, and can be computed using the large `NmzConeProperty` function. I added the placeholder to the `.gi` file now.

File: src/normaliz.cc [hunk 15]
```diff
@@ -54,6 +54,12 @@ using std::pair;
 using std::cerr;
 using std::endl;
 
+void signal_handler( int signal ){
+    libnormaliz::nmz_interrupted = true;
+}
+
+sighandler_t current_interpreter_sigint_handler;
```
Comment (sebasguts) [inline src/normaliz.cc]: Yes.

File: src/normaliz.cc (hunk 5 above)
Comment (sebasguts) [inline src/normaliz.cc]: There are two representations for Hilbert series in Normaliz. Currently, only the standard representation is returned.

Comment (fingolfin) [inline src/normaliz.cc]: That should then of course be added to the comment inside the file :-)

File: src/normaliz.cc (hunk 13 above)
Comment (sebasguts) [inline src/normaliz.cc]: Your assumption is completely right. This is btw. not really a good solution anyway. Copying the cone "detaches" it from `C`. Therefore, no further changes are propaged.

To really do it right, one should add some switch to tell the destructor `C` if the integer hull cone should be deleted along with `C`. Unfortunately, I could not think about a way that makes this possible :(.

Comment (fingolfin) [inline src/normaliz.cc]: I'd do it differently: add a "proxy cone" GAP wrapper object, similar to the existing cone wrapper; but this one in addition carries a (GAP) reference to the GAP wrapper of another cone... Then this "proxy cone" GAP wrapper would wrap `C->getIntegerHullCone()`, but in addition carries a reference to the GAP wrapper for `C`. As a result, the wrapper for `C` won't be garbage collected before the wrapper for the integer hull cone is (and hence `C` would be freed prematurely).

File: build-normaliz.sh [hunk 16]
```diff
@@ -59,7 +58,7 @@ else
     # So, we do not even try to figure out details, and instead
     # cross our fingers and hope for the best.
     #
-    # Expert user's can help with that, of course, by setting
+    # Expert user's can help with that, of courswe, by setting
```
Comment (fingolfin) [inline build-normaliz.sh]: New typo is introduced. Also an old typo: user's -> users

File: lib/cone_property_wrappers.gi [hunk 17]
```diff
@@ -1,63 +1,93 @@
-## This is an automatically generated file
+InstallGlobalFunction( NmzGenerators, cone -> NmzConeProperty(cone, "Generators" ) );
 
+InstallGlobalFunction( NmzExtremeRays, cone -> NmzConeProperty(cone, "ExtremeRays" ) );
 
-InstallGlobalFunction( NmzAffineDim, cone -> NmzConeProperty(cone, "AffineDim" ) );
+InstallGlobalFunction( NmzVerticesOfPolyhedron, cone -> NmzConeProperty(cone, "VerticesOfPolyhedron" ) );
```
Comment (fingolfin) [inline lib/cone_property_wrappers.gi]: How did you order these? In the old version, the properties were ordered alphabetically (possibly ignoring any `Is` prefix. Your order is different, leading to a bigger diff

File: src/normaliz.cc [hunk 18]
```diff
@@ -66,18 +97,37 @@ Obj NewCone(Cone<mpz_class>* C)
     return o;
 }
 
+Obj NewProxyCone(Cone<mpz_class>* C, Obj hull)
```
Comment (fingolfin) [inline src/normaliz.cc]: Why is the second argument named `hull`? Isn't this the parent of the hull? In any case, I'd name it `parent`, or `parentCone`.

File: src/normaliz.cc [hunk 19]
```diff
@@ -66,18 +97,37 @@ Obj NewCone(Cone<mpz_class>* C)
...
+
 /* Free function */
 void NormalizFreeFunc(Obj o)
 {
     delete GET_CONE<mpz_class>(o);
 }
 
+void NormalizProxyFreeFunc(Obj o)
+{
+    // Integer hull gets deleted when containing cone is deleted.
```
Comment (fingolfin) [inline src/normaliz.cc]: So if you use a single TNUM, then you have a single free func again, which could look like this:
```C++
#define IS_PROXY_CONE(o)  (SIZE_OBJ(o) == 2)
...
  if (!IS_PROXY_CONE(o))
    delete GET_CONE<mpz_class>(o);
  }
```

File: lib/normaliz.gd [hunk 20]
```diff
@@ -4,8 +4,10 @@
 NormalizObjectFamily := NewFamily( "NormalizObjectFamily" );
 
 DeclareCategory( "IsNormalizCone", IsObject );
+DeclareCategory( "IsNormalizIntHullCone", IsNormalizCone );
 
 BindGlobal("TheTypeNormalizCone", NewType( NormalizObjectFamily, IsNormalizCone ));
+BindGlobal("TheTypeNormalizIntHullCone", NewType( NormalizObjectFamily, IsNormalizIntHullCone ));
```
Comment (fingolfin) [inline lib/normaliz.gd]: Is there any particular reason why you want a new category and new type for the "proxy cones" / "IntHullCones" ?

File: src/normaliz.cc [hunk 21]
```diff
@@ -877,11 +954,19 @@ static Int InitKernel( StructInitInfo *module )
...
     InitMarkFuncBags(T_NORMALIZ, &MarkNoSubBags);
     InitFreeFuncBag(T_NORMALIZ, &NormalizFreeFunc);
-
+    
     CopyObjFuncs[ T_NORMALIZ ] = &NormalizCopyFunc;
     CleanObjFuncs[ T_NORMALIZ ] = &NormalizCleanFunc;
     IsMutableObjFuncs[ T_NORMALIZ ] = &NormalizIsMutableObjFuncs;
-
+    
+    InitMarkFuncBags(T_NORMALIZ_PROXY, &MarkTwoSubBags );
```
Comment (fingolfin) [inline src/normaliz.cc]: With a single TNUM, you can simply use `MarkAllSubBags` (for tiny overhead). Or write a custom mark function, like this:
```
static void NormalizMarkFunc(Obj o)
{
   if (IS_PROXY_CONE(o))
     MarkBag( ADDR_OBJ(o)[1] );
}

File: src/normaliz.h [hunk 22]
```diff
@@ -31,25 +31,37 @@ extern "C" {
...
 
 // old versions of libnormaliz (before 2.99.1) did not include such a define
 #ifndef NMZ_RELEASE
     static_assert(false,
-       "Your Normaliz version (unknown) is to old! Update to 3.0.0 or newer.");
+       "Your Normaliz version (unknown) is to old! Update to 3.3.0 or newer.");
 #endif
-#if NMZ_RELEASE < 30000
-    static_assert(false, "Your Normaliz version is to old! Update to 3.0.0 or newer.");
+#if NMZ_RELEASE < 30300
```
Comment (sebasguts) [inline src/normaliz.h]: I don't really know. It was like this before, I just updated the numbers ;). Changed it now.

File: lib/cone_property_wrappers.gi (hunk 17 above)
Comment (sebasguts) [inline lib/cone_property_wrappers.gi]: I honestly don't know anymore. Probably not at all. I now rearranged them.

**Comment by sebasguts** [comment]
Two questions here:
* I did the signal handling a bit different from how you suggested, using macros, so the signal handler change is more locally and the reset less implicit. However, it does not use sigaction atm, but I will change this.
* Should I rebase the commits as you suggested?

File: lib/cone_property_wrappers.gi (hunk 17 above)
Comment (fingolfin) [inline lib/cone_property_wrappers.gi]: Hmm, they still are shuffled in the overall diff...

File: src/normaliz.cc [hunk 23]
```diff
@@ -506,14 +644,16 @@ static Obj _NmzConePropertyImpl(Obj cone, Obj prop)
...
     // there is no ConeProperty HilbertQuasiPolynomial, it is part of the HilbertSeries
     // FIXME better way?
-    if (string(CSTR_STRING(prop)) == string("HilbertQuasiPolynomial")) {
-        C->compute(ConeProperties(libnormaliz::ConeProperty::HilbertSeries));
-        return NmzHilbertQuasiPolynomialToGAP(C->getHilbertSeries());
-    }
+//     if (string(CSTR_STRING(prop)) == string("HilbertQuasiPolynomial")) {
+//         C->compute(ConeProperties(libnormaliz::ConeProperty::HilbertSeries));
+//         return NmzHilbertQuasiPolynomialToGAP(C->getHilbertSeries());
+//     }
```
Comment (fingolfin) [inline src/normaliz.cc]: Why is this code commented out? Either remove it completely, or else add a comment as to why it is commented out?

File: src/normaliz.cc [hunk 24]
```diff
@@ -617,21 +766,79 @@ static Obj _NmzConePropertyImpl(Obj cone, Obj prop)
...
     case libnormaliz::ConeProperty::Sublattice:
-        C->compute(p);
-        return C->isComputed(p) ? True : False;
+        return _NmzBasisChangeIntern(C);
+    
+    case libnormaliz::ConeProperty::IntegerHull:
+    {
+        // The integer hull cone is copied because otherwise it would be deleted
+        // when C gets deleted.
+        // FIXME: Keep a reference
```
Comment (fingolfin) [inline src/normaliz.cc]: Is this FIXME still relevant? If so, I don't understand what it is about.

**Comment by fingolfin** [comment]
It is fine to use make it use sigaction in a future PR, no need to delay this one.

A cleaned up history of this PR would indeed be noce.

**Comment by fingolfin** [comment]
And I think we are pretty close to ready now, will take a full look again later.

**Comment by sebasguts** [comment]
So, except for a test for the `WeightedErhardPolynomial/Series` everything should be fixed now.

File: lib/cone_property_wrappers.gd [hunk 25]
```diff
@@ -42,6 +49,22 @@ DeclareGlobalFunction( "NmzDeg1Elements" );
...
+#! The embedding dimension is the dimension of the space in which the
+#! computation is done. It is the number of components of the output vectors.
+#! This value is always known directly after the creation of the cone.
+DeclareGlobalFunction("NmzEmbeddingDimension");
+
+#! @Arguments cone
+#! @Returns a matrix whose rows represent the equations
+#! @Description
+#! The equations cut out the linear space generated by the cone.
+#! Together with the support hyperplanes and the congruences it describes the
```
Comment (fingolfin) [inline lib/cone_property_wrappers.gd]: Grammar: "The equations" (plural) vs. "it describes" singular -- the latter "it" probably refers to the matrix returned by `NmzEquations`. Perhaps rephrase this as as follows (and then use the same sentence also for congruences and hyperplanes, which suffer from the same problem: "The equations, congruences and support hyperplanes together describe the lattice points of the cone."

File: lib/cone_property_wrappers.gd [hunk 26]
```diff
@@ -171,6 +200,14 @@ DeclareGlobalFunction( "NmzMultiplicity" );
 DeclareGlobalFunction( "NmzOriginalMonoidGenerators" );
 
 #! @Arguments cone
+#! @Returns the rank of the cone
+#! @Description
+#! The rank is the rank of the lattice generated by the lattice points of the cone.
```
Comment (fingolfin) [inline lib/cone_property_wrappers.gd]: Maybe start "The rank of the cone is" ?

Comment (fingolfin) [inline lib/cone_property_wrappers.gd]: Or, to match other entries: "This value is" ? I guess we can do that latter, but right now the documentation entries are a bit uneven in tone and expression, compared to each other.

File: lib/cone_property_wrappers.gd [hunk 27]
```diff
@@ -208,7 +245,10 @@ DeclareGlobalFunction( "NmzSupportHyperplanes" );
 #! @Arguments cone
 #! @Returns the triangulation
 #! @Description
-#! It is given as a list of pairs representing the maximal simplicial cones in the triangulation. The first pair entry is the key of the simplex, i.e. the indices of the generators with respect th the generators obtained by <Ref Func="NmzGenerators"/> (counting from 0). The second pair entry is the absolute value of the determinant of the generator matrix.
+#! It is given as a list of pairs representing the maximal simplicial cones in the triangulation.
```
Comment (fingolfin) [inline lib/cone_property_wrappers.gd]: "It is given" sounds really strange to me, and in addition I am no fan of potentially ambiguous references like that. Perhaps: "This returns a list of pairs pairs representing the maximal simplicial cones in a triangulation." (note the change of "the" to "a").

File: lib/cone_property_wrappers.gd [hunk 28]
```diff
@@ -208,7 +245,10 @@ DeclareGlobalFunction( "NmzSupportHyperplanes" );
...
+#! The first pair entry is the key of the simplex, i.e. the indices of the generators with respect
```
Comment (fingolfin) [inline lib/cone_property_wrappers.gd]: It took me a long time to parse that correctly. In particular, what is a "pair entry" ? After reading this again and again, I think you meant something like this: "The first entry of each pair in the list is [...]." and later "The second entry of each pair is [...]"

I also don't understand the explanation of what the "key of the simplex" is. Is it just an index? Or a list of indices? Make that clear, perhaps even be explicit: "... is a list of integers [a_1,...,a_n], which indicates [...]"

BTW, what exactly does `NmzGenerators` return? I didn't see it in our documentation.

And in how far does this list of pairs now indicate a "triangulation" (perhaps this is something trivial, and I just embarrassed myself with my lack of knowledge about triangulations; still, a hint here might help others, too)

File: lib/cone_property_wrappers.gd [hunk 29]
```diff
@@ -208,7 +245,10 @@ DeclareGlobalFunction( "NmzSupportHyperplanes" );
 #! @Arguments cone
 #! @Returns the triangulation
 #! @Description
-#! It is given as a list of pairs representing the maximal simplicial cones in the triangulation. The first pair entry is the key of the simplex, i.e. the indices of the generators with respect th the generators obtained by <Ref Func="NmzGenerators"/> (counting from 0). The second pair entry is the absolute value of the determinant of the generator matrix.
+#! It is given as a list of pairs representing the maximal simplicial cones in the triangulation.
+#! The first pair entry is the key of the simplex, i.e. the indices of the generators with respect 
+#! to the generators obtained by <Ref Func="NmzGenerators"/> (counting from 0).
```
Comment (fingolfin) [inline lib/cone_property_wrappers.gd]: It might also be nicer for the user if the indices matched the indices into `gens` in GAP, i.e. not starting from 0 but from 1.

File: lib/cone_property_wrappers.gd [hunk 30]
```diff
@@ -208,7 +245,10 @@ DeclareGlobalFunction( "NmzSupportHyperplanes" );
...
+#! The second pair entry is the absolute value of the determinant of the generator matrix.
```
Comment (fingolfin) [inline lib/cone_property_wrappers.gd]: "The second entry of each pair in the list is"

File: src/normaliz.cc [hunk 31]
```diff
@@ -350,8 +486,15 @@ static Obj _NmzConeIntern(Obj input_list)
             return Fail;
         }
         string type_str(CSTR_STRING(type));
-
         Obj M = ELM_PLIST(input_list, i+2);
+        if(type_str.compare("polynomial") == 0){
+            if(!IS_STRING_REP(M)){
```
Comment (fingolfin) [inline src/normaliz.cc]: Please add spaces suitably into the above two lines

File: src/normaliz.cc [hunk 32]
```diff
@@ -350,8 +486,15 @@ static Obj _NmzConeIntern(Obj input_list)
...
+                cerr << "Element " << i+2 << " of the input list must be a string" << endl;
+            }
+            polynomial = string(CSTR_STRING(M));
```
Comment (fingolfin) [inline src/normaliz.cc]: Uhm... so if we detect that `M` isn't a string, we print an error... and then access it as if it was a string anyway? Shouldn't we instead throw an exception?

**Comment by fingolfin** [comment]
Could you perhaps also cleanup the history? E.g. there are several commits dealing with proxy cones, which look as if they could be folded into a single one, namely 7c04300ae6c023ee58fc217c713fd5ca81dec9d8, af2cdb766d17221367a8d431d3f53acf9bb8e5e1 and f7013741bce4f82bac9e9e193666a590e04f4892 could be folded into d8ecc3806033364c191f51e8950dd60b9d5e3a88

**Comment by fingolfin** [comment]
Likewise, squash b59017d5f503119c2bb9dc2b4771c09b75f643b1 into 244bdf52bc82ae0b4c4190f2e190f917dfd082b3 ; and several more.

I could also do this if you prefer

File: lib/cone_property_wrappers.gd (hunk 28 above)
Comment (sebasguts) [inline lib/cone_property_wrappers.gd]: The documentation for `NmzGenerators` is there :). It returns the generators of the cone, i.e., the rays. I will nevertheless change this documentation.

File: build-normaliz.sh [hunk 33]
```diff
@@ -4,17 +4,17 @@
...
 #  ./configure --with-gaproot=GAPDIR
 #  make
 #
 
-if [ "$#" -ge 1 ]; then
-    GAPDIR=$1
-    shift
+if [ "$1" = "--gaproot" ]; then
+    GAPDIR=$2
+    shift 2;
```
Comment (fingolfin) [inline build-normaliz.sh]: Actually, this change breaks the `BuildPackages.sh` script bundled with GAP.

Which makes me wonder whether we really want that change? For what? If we want it, perhaps we can at least stay backwards compatible: if there is just one argument given, use that as GAPDIR?
